ambry-store/src/main/java/com/github/ambry/store/StatsBasedCompactionPolicy.java

```diff
@@ -121,23 +207,42 @@ private CostBenefitInfo getBestCandidateToCompact(NavigableMap<LogSegmentName, L
    */
   private CostBenefitInfo getCostBenefitInfo(LogSegmentName firstLogSegmentName, LogSegmentName lastLogSegmentName,
       NavigableMap<LogSegmentName, Long> validDataSizePerLogSegment, long segmentCapacity, long segmentHeaderSize,
-      long maxBlobSize) {
+      long maxBlobSize, boolean weightOnBenefit) {
     // cost = total valid data size (or data that needs to be copied over)
     // benefit = no of segments reclaimed = (total source segments - total target segments)
     // total target segments = Ceil (total cost / (segment capacity - segment header - max blob size))
     long totalCost = 0;
+    long adjustedTotalCost = 0;
     int totalSegmentsToBeCompacted = 0;
     LogSegmentName curSegmentName = firstLogSegmentName;
+    // weight more on benefits, don't let one single log segment with extreme small valid data leads to small compactions.
+    // Be default, the storeStatsBasedCompactionMinCostInPercentage is 0.06. 16G*0.06=1G
+    long minSegmentCost = storeConfig.storeStatsBasedCompactionMinCostInPercentage == 0 ? 0
+        : (long) (segmentCapacity * storeConfig.storeStatsBasedCompactionMinCostInPercentage);
     while (!curSegmentName.equals(lastLogSegmentName)) {
       totalCost += validDataSizePerLogSegment.get(curSegmentName);
+      adjustedTotalCost += validDataSizePerLogSegment.get(curSegmentName) < minSegmentCost ? minSegmentCost
+          : validDataSizePerLogSegment.get(curSegmentName);
+
       curSegmentName = validDataSizePerLogSegment.higherKey(curSegmentName);
       totalSegmentsToBeCompacted++;
     }
     // add the last log segment
     totalCost += validDataSizePerLogSegment.get(curSegmentName);
+    adjustedTotalCost += validDataSizePerLogSegment.get(curSegmentName) < minSegmentCost ? minSegmentCost
+        : validDataSizePerLogSegment.get(curSegmentName);
     totalSegmentsToBeCompacted++;
     long maxCapacityPerSegment = segmentCapacity - segmentHeaderSize - maxBlobSize;
     int benefit = totalSegmentsToBeCompacted - (int) Math.ceil(totalCost / (maxCapacityPerSegment * 1.0));
+
+    if (weightOnBenefit) {
+      if (totalCost < adjustedTotalCost) {
+        logger.info("getCostBenefitInfo {} cost {} adjusted cost {}",
+            validDataSizePerLogSegment.subMap(firstLogSegmentName, true, lastLogSegmentName, true).keySet(), totalCost,
+            adjustedTotalCost);
+        totalCost = adjustedTotalCost;
```

Comment:
Also, I still don't understand why calculating the `adjustedTotalCost` matters. I understand that we want to take everything that has lower than `StoreConfig.storeMaxLogSegmentValidDataPercentageToQualifyCompaction` valid data size. But why change their cost? What will happen if we use the `totalCost` instead of `adjustedTotalCost`?

Reply:
The reason is that our cost is too sensitive on the IO cost.
One extreme small valid size log segment will populate the data to make us pick just two log segments while all the neighbor log segments are already ready for compaction.
Taking the previous example,


If we set the threshold to 1GB, it means it worths to compact the log segments have less than 1GB valid size. 
In our current design, if we have log segments as below, (below is the valid size percentage). With current IO effort formula, we'll only compact log segment 3rd and 4th since their valid size is extreme small which is 1% and 1%.
It's cost is (0.01 + 0.01) * logSegmentSize / 1 = 0.02*logSegmentSize

double[] validPercentage = {0.9, 0.8, 0.05, 0.01, 0.01, 0.05, 0.1, 0.2, 0.4, 0.7}; 
But actually when the valid size percentage is less than some threshold, they are all ready to get compacted.

In the above example, if we set the threshold to 6%, with the adjustedTotalCost tuning, we'll compact log segment 2,3,4 and 5. The compaction benefit is 3, much better than the existing one which is 1. The IO cost added is very small.
The cost for the four log segment is (0.05+0.01+0.01+0.05) * logSegmentSize / 3 = 0.04*logSegmentSize.

You can see without the tuning, we'll pick log segment 3th and 4th only. But actually all 2, 3, 4 and 5 are ready for compaction. After the tuning, the cost of the group of 2 and 3 will increase. And then group of 2,3,4 and 5 will win.

New calculation for two log segments: (0.06+0.06) * logSegmentSize / 1 = 0.12 * logSegmentSize
New calculation for 4 log segments: (0.06 + 0.06 + 0.06+ 0.06) * logSegmentSize / 3 = 0.08 * logSegmentSize. So the second one wins.